### ISSUE TITLE
cgo_context_data is not resolved for the target platform

### ISSUE THREAD

**Issue by steeve** [bazelbuild/rules_go#2166]
# What version of rules_go are you using?

`master`

### What version of gazelle are you using?

`master`

### What version of Bazel are you using?

`0.28.0`

### Does this issue reproduce with the latest releases of all the above?

Yes.

### What operating system and processor architecture are you using?

`darwin_amd64`

### Any other potentially useful information about your toolchain?

Android NDK crosstool

### What did you do?

It seems that because `cgo_context_data` is resolved at toolchain creation time, the flags are the one from the host architecture, rather than the target.

I did a custom rule that depends on it and things are fine:
```python
load(
    "@bazel_tools//tools/cpp:toolchain_utils.bzl",
    "find_cpp_toolchain",
)
load(
    "@io_bazel_rules_go//go/private:providers.bzl",
    "CgoContextData",
)


def _test_cpp_impl(ctx):
    # cc_toolchain = find_cpp_toolchain(ctx)
    # print(">>>>", cc_toolchain)
    cgodata = ctx.attr._cgo_data[CgoContextData]
    print(">>>>>", cgodata)
    return []

test_cpp = rule(
    implementation = _test_cpp_impl,
    attrs = {
        "_cc_toolchain": attr.label(default = "@bazel_tools//tools/cpp:current_cc_toolchain"),
        "_cgo_data": attr.label(default = "@io_bazel_rules_go//:cgo_context_data"),
        "_xcode_config": attr.label(
            default = "@bazel_tools//tools/osx:current_xcode_config",
        ),
    },
    toolchains = ["@bazel_tools//tools/cpp:toolchain_type"],
    fragments = ["apple", "cpp"],
    provides = [CcInfo],
)
```

If I compile for android, everything in `cgo_context_data` is indeed for android. But if I build a `go_binary`, I see the content is from the host somehow.

My take on that is that the target is pulled before the platform split, and thus may still be referencing the host.

**Comment by steeve** [comment]
It seems the theory hold: if I make stdlib depend on the `@io_bazel_rules_go//:cgo_context_data` directly, then the flags are fine, while the ones the toolchain are not.

**Comment by steeve** [comment]
I'm thinking of just having a boolean in the toolchain info, and then somehow have go_context depend on cgo_context_data later on...

**Comment by steeve** [comment]
I have a patch that fixes it (basically makes `cgo_context_data` a dependency of `go_context_data`). I'll submit a PR tomorrow.

**Comment by steeve** [comment]
The WIP branch is at https://github.com/znly/rules_go/tree/steeve/toolchains

I'm mostly concerned about `cc_toolchain.all_files` which doesn't exist on bazel < 0.25, because ctx.files._cc_toolchain resolves the host toolchain, not the target (see 
 https://github.com/znly/rules_go/commit/c35ddff1fff0b55974ea34ad1e00f677dd9716d6#diff-bcfdf97a251cfcdf8dd613b5dd798c90R687)

**Comment by steeve** [comment]
I've added a helper in the compat package to deal with `cc_toolchain.all_files` https://github.com/znly/rules_go/commit/89c0120cdc64ba8bda816a714b2a8f0a26f5b598#diff-c5e9cbc666e381d81d591ba5fc56edc6

**Comment by jayconrod** [comment]
Hmm, this is a problem I'm not sure how to solve. I wonder if it could be as simple as adding `cfg = "target"` in the `cgo_context_data` attribute of `go_toolchain`? 

I don't actually have a cross-compiling C/C++ toolchain set up. I tried installing the Android NDK, but I couldn't get it to work with Bazel. Building with `--platforms=@io_bazel_rules_go//go/toolchain:android_arm64_cgo --extra_toolchains=@androidndk//:all` did not work: the toolchains don't appear to have been registered. Explicitly registering them in WORKSPACE fails with the message:

```
ERROR: While resolving toolchains for target //:hello: invalid registered toolchain '@androidndk//:aarch64-linux-android-clang8.0.7-libcpp': target does not provide the DeclaredToolchainInfo provider
```

Any idea what I'm doing wrong? I was following [Using the Android Native Development Kit with Bazel](https://docs.bazel.build/versions/master/android-ndk.html). I didn't install Android Studio or an SDK, so maybe the NDK doesn't work without an SDK.

---

I don't think your change in progress will work in all cases though. For the platforms declared in `@io_bazel_rules_go//go/toolchain`, we set `cgo_on` or `cgo_off` as one of the constraints, indicating whether a C toolchain is needed.

We can't assume people will use these platforms or these constraints, so we try to have reasonable default behavior whether neither constraint is specified. So when we declare cross-compiling Go toolchains, we declare a toolchain with the `cgo_on` constraint and a toolchain without either constraint. This means users "opt in" to requiring a C toolchain for cross-compilation. For the non-cross toolchains, this is reversed: one toolchain has `cgo_off` and the other has neither constraint. So users "opt out" of requiring a C toolchain.

@steeve Unfortunately, this means we can't write a `select` expression with `cgo_on` or `cgo_off`. If neither constraint is true, we can't infer anything: if we're cross-compiling, that means cgo is off; if we're not cross-compiling, that means cgo is on.

bazelbuild/bazel#8778 is a complicating factor (we can't safely set a default cgo constraint value), but because the default should depend on whether we're cross-compiling, I don't think resolving that issue would solve this.

**Comment by steeve** [comment]
I don't think the toolchain registrations for Android was released yet, I think it's only on `master`.

That said, you can do it manually (at least I am on 0.28.1):
```python
ANDROID_TOOLCHAINS = {
    "aarch64": "@androidndk//:aarch64-linux-android-clang7.0.2-libcpp",
    "arm": "@androidndk//:arm-linux-androideabi-clang7.0.2-v7a-libcpp",
    "x86_32": "@androidndk//:x86-clang7.0.2-libcpp",
    "x86_64": "@androidndk//:x86_64-clang7.0.2-libcpp",
}
[
    toolchain(
        name = "android_%s" % cpu,
        target_compatible_with = [
            "@platforms//os:android",
            "@platforms//cpu:%s" % cpu,
        ],
        toolchain = toolchain_,
        toolchain_type = "@bazel_tools//tools/cpp:toolchain_type",
    )
    for cpu, toolchain_ in ANDROID_TOOLCHAINS.items()
]
```
_________

As you mentioned in https://github.com/bazelbuild/bazel/issues/8778, we can't constrain on `cgo_on`, as that constraint isn't set by the C/C++ toolchain, which means it's discarded in the toolchain selection process because `@io_bazel_rules_go//go/toolchain:cgo_constaint` isn't set.

What I'm doing though, is expecting `rules_go` to set `cgo_off` on toolchains without cgo. Meaning that without the `cgo_off` constraint, a C/C++ toolchain is expected (because they will constrain on only the os/arch).
You can see that at https://github.com/znly/rules_go/commit/89c0120cdc64ba8bda816a714b2a8f0a26f5b598#diff-0ea22445d900db774c8deb00f4a4df55R82-R85

I think this solution circumvents the problem entirely.

______

Good call about the `cfg = "target"`. That said, it doesn't bother me to have `go_context_data` depend on `cgo_context_data` directly, imho.

**Comment by jayconrod** [comment]
> What I'm doing though, is expecting rules_go to set cgo_off on toolchains without cgo. Meaning that without the cgo_off constraint, a C/C++ toolchain is expected (because they will constrain on only the os/arch).

That would break people who are defining their own `platforms` without requiring `cgo_off` or `cgo_on` explicitly. People would always have to set one or the other. Depending on how this is set up, it would also be a problem for the default target platform (no `--platforms`), which doesn't require those constraints either.

**Comment by steeve** [comment]
That is a very valid point. That said, if they do define their own platform, how will they define their potential dependency (or not) to a C/C++ toolchain ?

**Comment by jayconrod** [comment]
> That said, if they do define their own platform, how will they define their potential dependency (or not) to a C/C++ toolchain ?

Currently, for every `goos` / `goarch`, there's a `go_toolchain` that has neither `cgo_on` nor `cgo_off` that will be selected. For the host platform, it has a dependency on `cgo_context_data`. For the target platform, it does not (only the `cgo_on` toolchains have that dependency).

> That said, with my solution if they don't define one (cgo_off in my case), it assumes they do have a C/C++ toolchain for the target platform.

Assuming `cfg = "target"` does not work (I'm still hoping it does), this may be the only solution. This would only affect people defining their own platforms, which is a pretty high level of sophistication, so I think it's a reasonable change. However, it is a breaking change, and we wouldn't be able to backport to the `release-0.19` branch.

**Comment by steeve** [comment]
Agreed about the sophistication part.
That said, and forgive me I'm wrong, but don't they already need to define `cgo_on` or `cgo_off` ? With my change, they would only need to define `cgo_off` if they don't want a dependency on a C++ toolchain.

I'm completely fine with waiting for 0.20. That said, 0.19 is not working for me in a cross compile scenario, so I would argue a release would be useful for those with that use case (are there anyone besides me ? :D)

But I'd wager (looking at twitter a while ago) a lot of people were bitten by 0.26/0.27, so perhaps they haven't upgraded yet?

I'll submit the PR and let's discuss there. I'll submit a test project too so you can experiment, I'm sure we'll figure out what works best.

**Comment by jayconrod** [comment]
> That said, and forgive me I'm wrong, but don't they already need to define cgo_on or cgo_off ? 

For cross-compilation, users only need to depend on `cgo_on` if they want to use a C/C++ toolchain. If they don't depend on either value, they'll get a toolchain without the C/C++ dependency which doesn't require either. The code you linked is what controls that, specifically L117-119.

**Comment by jayconrod** [comment]
Before moving forward with a PR, I'd like to confirm that `cfg = "target"` does not work. I've consistently failed to set up the Android NDK toolchain, and I wonder if you can tell me what I'm doing wrong. Here's what I'm doing:

* Using Bazel 0.29.0rc3 on darwin, amd64 with rules_go at master.
* Downloaded Android NDK r20 from https://developer.android.com/ndk/downloads and extracted to /opt/android-ndk-r20.
* Have the following workspace:

```
-- WORKSPACE --
local_repository(
    name = "io_bazel_rules_go",
    path = "/Users/jayconrod/go/src/github.com/bazelbuild/rules_go",
)

load("@io_bazel_rules_go//go:deps.bzl", "go_register_toolchains", "go_rules_dependencies")

go_rules_dependencies()

go_register_toolchains()

android_ndk_repository(
    name = "androidndk",
    path = "/opt/android-ndk-r20",
)
-- BUILD.bazel --
ANDROID_TOOLCHAINS = {
    "aarch64": "@androidndk//:aarch64-linux-android-clang8.0.7-libcpp",
    # "aarch64": "@androidndk//:aarch64-linux-android-clang7.0.2-libcpp",
    # "arm": "@androidndk//:arm-linux-androideabi-clang7.0.2-v7a-libcpp",
    # "x86_32": "@androidndk//:x86-clang7.0.2-libcpp",
    # "x86_64": "@androidndk//:x86_64-clang7.0.2-libcpp",
}

[
    toolchain(
        name = "android_%s" % cpu,
        target_compatible_with = [
            "@bazel_tools//platforms:android",
            "@bazel_tools//platforms:" + cpu,
        ],
        toolchain = toolchain_,
        toolchain_type = "@bazel_tools//tools/cpp:toolchain_type",
    )
    for cpu, toolchain_ in ANDROID_TOOLCHAINS.items()
]

cc_binary(
    name = "hello",
    srcs = ["hello.c"],
)
-- hello.c --
#include <stdio.h>

int main() {
  printf("hello\n");
  return 0;
}
```

* Building with this command:

```
bazel build --platforms=@io_bazel_rules_go//go/toolchain:android_arm64 --extra_toolchains=@//:all --toolchain_resolution_debug :hello
```

From the output, it looks like the correct toolchain is selected, but it isn't actually used to compile C/C++ code. The output binary is still Mach-O x86_64. Is there a flag I'm missing there?

**Comment by steeve** [comment]
Hum that's weird indeed:
```
-- WORKSPACE --
android_ndk_repository(
    name = "androidndk",
    api_level = 21,
)
 # reset to current master
local_repository(
    name = "io_bazel_rules_go",
    path = "../../znly/rules_go",
)

load("@io_bazel_rules_go//go:deps.bzl", "go_rules_dependencies", "go_register_toolchains", "go_download_sdk")
go_rules_dependencies()

go_register_toolchains()

-- BUILD.bazel --
cc_binary(
    name = "main_cc",
    srcs = ["main.c"],
)

ANDROID_TOOLCHAINS = {
    "aarch64": "@androidndk//:aarch64-linux-android-clang7.0.2-libcpp",
    "arm": "@androidndk//:arm-linux-androideabi-clang7.0.2-v7a-libcpp",
    "x86_32": "@androidndk//:x86-clang7.0.2-libcpp",
    "x86_64": "@androidndk//:x86_64-clang7.0.2-libcpp",
}

[
    toolchain(
        name = "android_%s" % cpu,
        target_compatible_with = [
            "@bazel_tools//platforms:android",
            "@bazel_tools//platforms:%s" % cpu,
        ],
        toolchain = toolchain_,
        toolchain_type = "@bazel_tools//tools/cpp:toolchain_type",
    )
    for cpu, toolchain_ in ANDROID_TOOLCHAINS.items()
]
```

Then:
```
$ bazel build --platforms=@io_bazel_rules_go//go/toolchain:android_arm64 //:main_cc
Target //:main_cc up-to-date:
  bazel-bin/main_cc
$ file bazel-bin/main_cc
bazel-bin/main_cc: ELF 64-bit LSB pie executable ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /system/, with debug_info, not stripped
```

**Comment by steeve** [comment]
I just tried `cfg = "target"`:
```python
        "cgo_context_data": attr.label(
            providers = [CgoContextData],
            cfg = "target",
            doc = "A target that collects information about the C/C++ toolchain.",
        ),
```

Unfortunately it doesn't work, it still returns the host compilation options :(

**Comment by steeve** [comment]
I think I see the problem with my method: only specifying `os`/`arch` constraints without `cgo_off` will assume CGo, where it's not the case currently.

The problem is the same on the other side, however: specifying `cgo_on` (reversing the problem) discards the C/C++ toolchain because it doesn't declare that constraint class (toolchain debug says it's missing the `cgo_constraint`.

**Comment by jayconrod** [comment]
Thanks for that example. I got it to cross-compile. I think I was missing `--incompatible_enable_cc_toolchain_resolution`. Not sure why I didn't see that earlier.

Confirmed that `cfg = "target"` still gives something in the host configuration.

I'll ask some folks working on Bazel configurability if there's another way around this.

**Comment by jayconrod** [comment]
> The problem is the same on the other side, however: specifying cgo_on (reversing the problem) discards the C/C++ toolchain because it doesn't declare that constraint class (toolchain debug says it's missing the cgo_constraint.

This should be okay. When a `toolchain` is declared, in its `target_compatible_with` (or `exec_compatible_with`) attribute, if none of the `constraint_values` for a particular `constraint_setting` are declared, then that setting doesn't affect selection (assuming there's no default value). So if a toolchain specifies neither `cgo_on` nor `cgo_off`, either value in the `platform` is okay for that toolchain (or no value).

**Comment by steeve** [comment]
Thinking out loud, and looking at https://github.com/znly/rules_go/commit/a1ec9684da5ef57bc6bd2ff0375987ec8a0eeb8d#diff-0ea22445d900db774c8deb00f4a4df55R82-R85, I'm wondering why we bother having different toolchains for CGo.

Since the dependency is expressed here, we could very well have CGo enabled via something completely different, like `--define=io.bazel.rules.go.cgo=1` and be done with it? The whole CGo enabled for host but not cross compilation makes it difficult via `select`...

**Comment by steeve** [comment]
> This should be okay. When a toolchain is declared, in its target_compatible_with (or exec_compatible_with) attribute, if none of the constraint_values for a particular constraint_setting are declared, then that setting doesn't affect selection (assuming there's no default value). So if a toolchain specifies neither cgo_on nor cgo_off, either value in the platform is okay for that toolchain (or no value).
Indeed, I just confirmed it!

**Comment by jayconrod** [comment]
> Thinking out loud, and looking at znly@a1ec968#diff-0ea22445d900db774c8deb00f4a4df55R82-R85, I'm wondering why we bother having different toolchains for CGo.

If the toolchain doesn't depend on `cgo_context_data`, we don't need cgo and non-cgo variants. We didn't have them before 0.19.0. I thought it was going to solve this problem, but unfortunately not.

> Since the C/C++ dependency is expressed here, we could very well have CGo enabled via something completely different, like --define=io.bazel.rules_go.cgo=1 and be done with it? The whole CGo enabled for host but not cross compilation makes it difficult via select..

We could do something with build settings and configuration transitions. But I don't think it would end up being all that different than controlling cgo with `platform` and `constraint_values`.

> Also, one could argue that having a C/C++ toolchain shouldn't be required to build Go programs, even for the host? Perhaps bazel itself requires it though (I know proto rules do)

It's required for the host, at least by default. People would be *very* unhappy if they had to pass an extra flag to make cgo work.

**Comment by steeve** [comment]
Another solution would be to have two `go_context_data`:
- one without the dep to `cgo_context_data`
- one with it

And control wether a `go_libray` depends on which via `cgo = True` in a macro ?

**Comment by steeve** [comment]
I managed to make it work in two ways:
- via a transition/config_setting that triggers on `cgo = True` and makes `go_context_data` depend or not on `cgo_context_data`. Downside is that the output path now has a transition slug:
```
Target //:test up-to-date:
  bazel-out/darwin-fastbuild-ce84244e6d5a0d2c885fca1c324640b0/bin/darwin_amd64/test
```
- by duplicating `go_context_data` with one having ` dependency on `cgo_context_data` and one which doesn't, then depend on whichever in the macro. This method is probably more portable across bazel versions

With both methods, the downside is that I had to get rid of `pure` mode because the graph now has both pure and non-pure nodes in it. If we could find a way to propagate `pure = off` in the graph should one node be `cgo = True`, perhaps that would fix it. There is also the problem of `_stdlib`.

This works well, and only requires a C/C++

**Comment by jayconrod** [comment]
I thought about this a bit more and chatted with @katre.

### Long term plan

We should declare one `go_toolchain` per `go_sdk`. Right now, we declare a `go_toolchain` per supported platform with variants for cgo enabled or not. The purpose of this is so we can configure the toolchain via attributes, as we are doing now with `cgo_context_data`.

We should instead configure using build settings. We would have build settings for `goos`, `goarch`, `pure`, `static`, and other stuff currently found in mode. Defaults can be extracted from the target platform with `select` and `constraint_values`. Most of these could be set on the  command line, like we do with `--features=pure` now (which is a hack).

Since we're using build settings, we could drop our archive aspect and implement `goos`, `goarch` and other attributes of `go_binary` using an incoming transition.

We could use an incoming `transition` and `config_setting` in `go_context_data` to depend on `cgo_context_data` or not.

Most of this work was already planned. @juliexxia may look at this, but she's temporarily on another project for a few months. One of us will implement this eventually, depending on how the timing works out.

As of Bazel 0.29.0rc3 though, the whole API is marked experimental, and there's a whitelist for using `transition` (which I think requires a custom built Bazel binary). So I'm not confident moving forward with this yet. @gregestren What are the steps to get this stabilized? Would we need a proof of concept PR in rules_go first?

### Short term plan

@steeve This is pretty much what you proposed earlier. The dependency on `cgo_context_data` should be moved to `go_context_data`, gated by a `select`. If the `cgo_off` contraint value is present, `cgo_context_data` will not be required. Otherwise, it will be.

This means that the default target platform will require cgo. All the platforms defined in `@io_bazel_rules_go//go/toolchain` will have either `cgo_on` or `cgo_off` set, so users targeting those will not be broken. However, anyone defining their own `platform` without specifying `cgo_off` may suddenly have a cgo dependency.

I don't think there's a way to fix this without breaking anyone. This will be called out in the notes for the next release. People defining their own `platform` are pretty sophisticated, and this already wasn't working correctly, so hopefully it won't be too disruptive.

I'll put together a PR for this shortly.

**Comment by gregestren** [comment]
Using transitions is non-experimental and while there's technically a whitelist requirement the Bazel whitelist is open to everything. So there are no practical blockers to using these mechanisms. This should all be true in release Bazel.

**Comment by jayconrod** [comment]
@gregestren and I chatted offline. He pointed me to a [useful example](https://github.com/gregestren/snippets/blob/master/cc_binary_selectable_copts/defs.bzl#L56-L65). Currently, rules using transitions need to define a `_whitelist_function_transition` attribute that points to `@bazel_tools//tools/whitelists/function_transition_whitelist`. However, the Bazel version of the whitelist has been expanded to include everything.

There are a number of documentation pages indicating transitions are experimental, but this only applies to split transitions (which rules_go will not need). The rest of the API is stable.

So with that clarification, I don't believe anything blocks migrating to build settings and transitions. Greg, Julie, or I will take a look at this, depending on available bandwidth.

For now though, I'll move forward with the short term plan outlined above.

**Comment by steeve** [comment]
> We should declare one `go_toolchain` per `go_sdk`. Right now, we declare a `go_toolchain` per supported platform with variants for cgo enabled or not. The purpose of this is so we can configure the toolchain via attributes, as we are doing now with `cgo_context_data`.

That looks great!

> We should instead configure using build settings. We would have build settings for `goos`, `goarch`, `pure`, `static`, and other stuff currently found in mode. Defaults can be extracted from the target platform with `select` and `constraint_values`. Most of these could be set on the command line, like we do with `--features=pure` now (which is a hack).

Not sure about this one, although I kind of agree. The thing is, I find it quite elegant that a particular GOOS/GOARCH be properly set by a platform (perhaps even constrained by others, such as a CC toolchain).

> Since we're using build settings, we could drop our archive aspect and implement `goos`, `goarch` and other attributes of `go_binary` using an incoming transition.
> 
> We could use an incoming `transition` and `config_setting` in `go_context_data` to depend on `cgo_context_data` or not.

That looks great too.

> As of Bazel 0.29.0rc3 though, the whole API is marked experimental, and there's a whitelist for using `transition` (which I think requires a custom built Bazel binary). So I'm not confident moving forward with this yet. @gregestren What are the steps to get this stabilized? Would we need a proof of concept PR in rules_go first?

> @gregestren and I chatted offline. He pointed me to a useful example. Currently, rules using transitions need to define a _whitelist_function_transition attribute that points to @bazel_tools//tools/whitelists/function_transition_whitelist. However, the Bazel version of the whitelist has been expanded to include everything.

I saw people using it as far as January this year (I think it was rules_docker). I myself am using it in my soon-to-be open sourced `rules_gomobile`. At least it was available in Bazel 0.23. That said, I _think_ the `@bazel_tools` approach does require https://github.com/bazelbuild/bazel/commit/a33545b8b8cf2a9fff06b0ad28f97e1671962ba5, but I'm not sure. Else people would need to create a whitelist file in their `WORKSPACE`. That might need some checking out.

> People defining their own `platform` are pretty sophisticated, and this already wasn't working correctly, so hopefully it won't be too disruptive.

Yeah I think this is a good compromise.

**Comment by jayconrod** [comment]
> The thing is, I find it quite elegant that a particular GOOS/GOARCH be properly set by a platform (perhaps even constrained by others, such as a CC toolchain).

I *think* this should still work. We could set the default value of `goos`, `goarch` build settings using `select` on whatever `constraint_values` are in the target platform.
